### PR TITLE
PHASE-I self service Monitoring

### PR DESCRIPTION
**Pull request by avdhoot**

### CODE REVIEW

File: pkg/apis/addtoscheme_monitoring_v1beta1.go [hunk 1]
```diff
@@ -0,0 +1,26 @@
+/*
+Copyright 2018-2019 Ridecell, Inc.
```
Comment (coderanger) [inline pkg/apis/addtoscheme_monitoring_v1beta1.go]: It's not 2018 :)

File: pkg/apis/monitoring/v1beta1/alertmanagerconfig_types.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// AlertManagerConfigSpec defines the desired state of AlertManagerConfig
+type AlertManagerConfigSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+	Data                  map[string]string `json:"data,omitempty"`
+	AlertManagerName      string            `json:"alertManagerName,omitempty"`
+	AlertManagerNamespace string            `json:"alertMangerNamespace,omitempty,"`
```
Comment (coderanger) [inline pkg/apis/monitoring/v1beta1/alertmanagerconfig_types.go]: Extra trailing `,`

File: pkg/apis/monitoring/v1beta1/monitor_types.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	// pov1 "github.com/coreos/prometheus-operator/pkg/apis/monitoring/v1"
```
Comment (coderanger) [inline pkg/apis/monitoring/v1beta1/monitor_types.go]: Comment code, can remove.

File: pkg/apis/monitoring/v1beta1/monitor_types.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// MonitorSpec defines the desired state of Monitor
+type MonitorSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+	Notify           Notify            `json:"notify,omitempty"`
+	MetricAlertRules []MetricAlertRule `json:"metricsAlerts,omitempty"`
```
Comment (coderanger) [inline pkg/apis/monitoring/v1beta1/monitor_types.go]: It's generally good to keep the JSON name similar to the field name, so `metricAlertRules`.

File: pkg/apis/monitoring/v1beta1/v1beta1_suite_test.go [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	"github.com/Ridecell/ridecell-operator/pkg/test_helpers"
+)
+
+var testHelpers *test_helpers.TestHelpers
+
+func TestTemplates(t *testing.T) {
```
Comment (coderanger) [inline pkg/apis/monitoring/v1beta1/v1beta1_suite_test.go]: Not templates tests :) (doesn't actually matter as long as the name starts with `Test` though)

File: pkg/apis/monitoring/v1beta1/v1beta1_suite_test.go [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "AlertManagerConfig types")
+	RunSpecs(t, "Monitor types")
```
Comment (coderanger) [inline pkg/apis/monitoring/v1beta1/v1beta1_suite_test.go]: You only want one of these, it's the name of the suite, so `monitoring v1beta1 types`.

File: pkg/controller/alertmanagerconfig/components/alertmanagerconfig.go [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+		if config.Spec.AlertManagerName == instance.Spec.AlertManagerName {
+			route, _ = base64.StdEncoding.DecodeString(config.Spec.Data["routes"])
+			receiver, _ = base64.StdEncoding.DecodeString(config.Spec.Data["receiver"])
+			inhibitRule, _ = base64.StdEncoding.DecodeString(config.Spec.Data["inhibitRule"])
+			errRo := yaml.Unmarshal(route, routetype)
+			errRe := yaml.Unmarshal(receiver, receivertype)
+			errIn := yaml.Unmarshal(inhibitRule, inhibitRuletype)
+			if errRo != nil || errRe != nil || errIn != nil {
+				glog.Errorf("failed to load yaml for %s in %s", config.Name, config.Namespace)
+				continue
```
Comment (coderanger) [inline pkg/controller/alertmanagerconfig/components/alertmanagerconfig.go]: I don't think we should ignore parsing errors?

File: pkg/controller/alertmanagerconfig/components/alertmanagerconfig.go [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+		return components.Result{}, errors.Wrapf(err, "failed to list alermanagerconfig.")
+	}
+	// Merge  all AlertManagerConfig's
+	for _, config := range alertList.Items {
+		var route, receiver, inhibitRule []byte
+		routetype := &alertconfig.Route{}
+		receivertype := &alertconfig.Receiver{}
+		inhibitRuletype := &alertconfig.InhibitRule{}
+		if config.Spec.AlertManagerName == instance.Spec.AlertManagerName {
+			route, _ = base64.StdEncoding.DecodeString(config.Spec.Data["routes"])
```
Comment (coderanger) [inline pkg/controller/alertmanagerconfig/components/alertmanagerconfig.go]: Shouldn't ignore errors here either.

File: pkg/controller/monitor/components/logrule.go [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+
+func (_ *logruleComponent) IsReconcilable(_ *components.ComponentContext) bool {
+	return true
+}
+
+func (comp *logruleComponent) Reconcile(ctx *components.ComponentContext) (components.Result, error) {
+	fmt.Println("in logrule")
+	instance := ctx.Top.(*monitoringv1beta1.Monitor)
+	rules, _ := yaml.Marshal(instance.Spec.LogAlertRules)
+	fmt.Println("logrules", string(rules))
```
Comment (coderanger) [inline pkg/controller/monitor/components/logrule.go]: Debugging printf.

File: pkg/controller/monitor/components/logrule.go [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+	// get folderid
+	return components.Result{}, nil
+}
+
+func sumoRequest(url, method string, body []byte) ([]byte, error) {
+	client := &http.Client{}
+	req, _ := http.NewRequest(method, baseURL+url, bytes.NewBuffer(body))
+	req.SetBasicAuth("su5QkKDVw7s2AF", "i6KfitiTGzXlHXdtPDgBHrS3D0Dabogs6g6zeYOpsangmb9q1foXEbGZyijJa1dz")
```
Comment (coderanger) [inline pkg/controller/monitor/components/logrule.go]: ☠️

File: pkg/controller/monitor/components/monitor.1.old [hunk 11]
```diff
@@ -0,0 +1,113 @@
+/*
```
Comment (coderanger) [inline pkg/controller/monitor/components/monitor.1.old]: I don't think you meant to commit this file.

File: pkg/controller/alertmanagerconfig/components/alertmanagerconfig.go (hunk 7 above)
Comment (avdhoot) [reply pkg/controller/alertmanagerconfig/components/alertmanagerconfig.go]: It Is required to find faulty AlertManagerConfig.

Comment (avdhoot) [reply pkg/controller/alertmanagerconfig/components/alertmanagerconfig.go]: I miss read what you trying to say. But I will set object.status to ERROR

Comment (coderanger) [reply pkg/controller/alertmanagerconfig/components/alertmanagerconfig.go]:
```go
if instance.UUID == config.UUID {
  return components.Result{}, errors.Wrapf(err, "error parsing yaml")
} else {
  continue
}
```

File: pkg/controller/alertmanagerconfig/components/alertmanagerconfig.go [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+		if config.Spec.AlertManagerName == instance.Spec.AlertManagerName {
+			errRo := yaml.Unmarshal([]byte(config.Spec.Data["routes"]), routetype)
+			errRe := yaml.Unmarshal([]byte(config.Spec.Data["receiver"]), receivertype)
+			errIn := yaml.Unmarshal([]byte(config.Spec.Data["receiver"]), inhibitRuletype)
+			if errRo != nil || errRe != nil || errIn != nil {
+				glog.Errorf("failed to load yaml for %s in %s", config.Name, config.Namespace)
+				if instance.UID == config.UID {
+					return components.Result{}, errors.Errorf("failed to load yaml for %s in %s", config.Name, config.Namespace)
+				}
+			}
```
Comment (coderanger) [inline pkg/controller/alertmanagerconfig/components/alertmanagerconfig.go]: Either need to put back the `continue` or put all the appends below in an `else`.

**Pull request merged**